themes-default/slim/test/specs/quality-pill.spec.js

```diff
@@ -16,72 +16,101 @@ test.beforeEach(t => {
     t.context.store = new Store({ state });
 });
 
-test('1080p WEB-DL', t => {
+/**
+ * Generate a simple test case
+ *
+ * @param {*} t - AVA test
+ * @param {string} message - Test message
+ * @param {object} propsData - Props to pass to the component.
+ */
+const pillTestCase = (t, message, propsData) => {
     const { localVue, store } = t.context;
-    const wrapper = mount(QualityPill, {
+
+    t.snapshot(mount(QualityPill, {
         localVue,
         store,
-        propsData: {
-            quality: 128
-        }
+        propsData
+    }).html(), message);
+};
+
+test('renders quality pills correctly', t => {
+    pillTestCase(t, 'No qualities', {
```

Comment:
This would come in handy in quite a few places. Why don't we move this to a helper and make it a higher level function?

```js
const generatePropTest = component => (t, message, propsData) => {
    const { localVue, store } = t.context;

    t.snapshot(mount(component, {
        localVue,
        store,
        propsData
    }).html(), message);
};
const pillTestCase = generatePropTest(QualityPill);
```

Reply:
That's a good point.

1. I'm not so sure how it should be structured tree wise though, I mean, in `test/helpers/[what here?]`.
2. I can't find a way of documenting the inner function using JSDoc without assigning it to an inner variable.

This seemed to work best:
```js
/**
 * Generate a simple component prop test case.
 *
 * @param {*} component - A Vue component.
 * @returns {propTest} Test case function.
 */
const generatePropTest = component => {
    /**
     * A simple component prop test case.
     *
     * @param {*} t - AVA test with a context containing `localVue` and `store` properties.
     * @param {string} message - Test message.
     * @param {object} propsData - Props to pass to the component.
     */
    const propTest = (t, message, propsData) => {
        const { localVue, store } = t.context;

        t.snapshot(mount(component, {
            localVue,
            store,
            propsData
        }).html(), message);
    };
    return propTest;
};
```

Reply:
Aren't the tests failing early as we have the `failFast` flag enabled?

Reply:
`propTest` may be better outside of that function so the js-doc resolution works properly.

Reply:
> `propTest` may be better outside of that function so the js-doc resolution works properly.

If I put `propTest` outside the function I can't reference `component`. AFAIK it must be an inner function to work.

> Aren't the tests failing early as we have the `failFast` flag enabled?

I'll check that.

> 1. I'm not so sure how it should be structured tree wise though, I mean, in `test/helpers/[what here?]`.

Can you please answer that? I don't know where to put the function. :confused:

Reply:
`failFast` flag doesn't affect this scenario. If one assertion fails inside a `test`, it fails the whole test.
I'm not saying that's wrong, I'm just saying we may want to have it be multiple `test`s, instead of one `test` with multiple assertions.

Reply:
> If I put propTest outside the function I can't reference component. AFAIK it must be an inner function to work.

@sharkykh change `const propTest =` to `return`. That'll be the same as what I was trying to achieve.

> I'm not saying that's wrong, I'm just saying we may want to have it be multiple tests, instead of one test with multiple assertions.

We could generate the test as well with that function, these are called [test macros](https://github.com/avajs/ava#test-macros).

Reply:
jsdoc doesn't recognize the return arrow function as a function. that's why I used a variable.
I said it in my first reply:
>I can't find a way of documenting the inner function using JSDoc without assigning it to an inner variable.

It just didn't acknowledge the fact the wrapper function returns a function with 3 arguments.

Reply:
Curried functions do work with jsdoc, see [this](https://github.com/jsdoc3/jsdoc/issues/1286#issuecomment-341911947).

Reply:
It's not perfect, as you can't document the parameters of the inner function:
![image](https://user-images.githubusercontent.com/10238474/45698458-b1e36080-bb70-11e8-8ae0-1d36d989f9e6.png)
:disappointed:

Reply:
So we'll go with `#2` then?

Reply:
yeah I guess...

and `#2` is the current code